tools/create-upscale-inventory.py

```diff
@@ -0,0 +1,298 @@
+    args = parser.parse_args()
+    return args
+
+
+def read_hosts_from_inventory(inventory_file):
+    """
+    Ansible inventory isn't exactly the same as an INI file. You could use a
+    ConfigParser, but it doesn't quite give you the right output. However, you
+    can use the Ansible inventory manager.
+    http://docs.ansible.com/ansible/latest/dev_guide/developing_api.html
+    https://github.com/ansible/ansible/blob/devel/lib/ansible/inventory
+
+    Update: Ansible's Inventory Manager is not useful for WRITING, but the
+    ConfigParser can be made to provide the right output
+
+    Args:
+        inventory_file: the filename containing the inventory.
+
+    Returns:
+        a dict representing the hosts from the inventory file.
+        The following keys should be present:
+        ['all', 'dns', 'loadbalancers', 'masters', 'ungrouped', 'etcd',
+         'OSEv3', 'nodes']
+    """
+    ivm = InventoryManager(loader=DataLoader(), sources=[inventory_file])
+    return ivm.get_groups_dict()
+
+
+def read_hosts_from_config_yaml(vars_config_file):
+    """
+    Read from the ansible vars file group_vars/all.yml
+
+    Return a list of tuples, each being (hostname, ip_address)
+
+    """
+    with open(vars_config_file, 'r') as vars_fh:
+        vars_config = yaml.load(vars_fh)
+
+    output = []
+    for k, v in vars_config.items():
+        if k in ['haproxy_details', 'master_details', 'worker_details']:
+            for ip, host in v.items():
+                output.append((host, ip))
+    return output
+
+
+def find_missing_hosts(all_inventory_hosts, new_hosts_list):
+    """
+    Look for any hostnames that are only in one of the two
+    config files
+
+    Args:
+        all_inventory_hosts: a list of hostnames, originating
+            from the inventory file (
+        new_hosts_list: a list of tuples of all the hosts that should be
+            in the enlarged cluster. Each is (hostname, ip_address)
+    """
+
+    # Does every host in all.yml exist in the inventory?
+    discovered_hosts = []
+    added = []
+
+    for hostname, ip_addr in new_hosts_list:
+        discovered_hosts.append(hostname)
+        if hostname not in all_inventory_hosts:
+            added.append((hostname, ip_addr))
+
+    return added
+
+
+def write_scale_up_inventory(old_inventory_file, upscaled_inventory_file,
+                             new_hosts, debug=False):
+    """
+    Scripted implementation of the instructions at
+    https://docs.openshift.com/container-platform/3.5/install_config/
+        adding_hosts_to_existing_cluster.html#adding-nodes-advanced
+
+    """
+    inventory = AnsibleInventory(filename=old_inventory_file)
+
+    for host, ip_arg in new_hosts:
+        # the new host is either "master-*.openstacklocal" or
+        # worker-*.openstacklocal"
+
+        host_type = ''
+        host_num = -1
+        if debug:
+            print("DEBG: {}".format(host))
+        mtch = re.match('(worker|master)-(\d+).openstacklocal', host)
```

Comment:
it's not always openstacklocal - we could read the precise hostname from the config